gapis/api/transform2/transformer.go

```diff
@@ -0,0 +1,108 @@
+// Copyright (C) 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transform2
+
+import (
+	"context"
+
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/gapis/api"
+	"github.com/google/gapid/gapis/api/command_generator"
+	"github.com/google/gapid/gapis/replay/builder"
+)
+
+type Transformer struct {
+	transforms       []Transform
+	commandGenerator command_generator.CommandGenerator
+	state            *api.GlobalState
+	builder          *builder.Builder
+}
+
+func NewTransformer(commandGenerator command_generator.CommandGenerator, state *api.GlobalState, builder *builder.Builder) Transformer {
+	return Transformer{
+		transforms:       make([]Transform, 0),
+		commandGenerator: commandGenerator,
+		state:            state,
+		builder:          builder,
+	}
+}
+
+func (transformer *Transformer) AddTransform(transforms ...Transform) {
+	transformer.transforms = append(transformer.transforms, transforms...)
+}
+
+func (transformer *Transformer) buildCmd(ctx context.Context, id api.CmdID, cmds []api.Cmd) error {
+	for i, cmd := range cmds {
+		transformer.builder.BeginCommand(uint64(id), cmd.Thread())
+		err := cmd.Mutate(ctx, id, transformer.state, transformer.builder, nil)
+		if err == nil {
+			transformer.builder.CommitCommand()
+		} else {
+			transformer.builder.RevertCommand(err)
+			log.W(ctx, "Failed to write command [%v:%V] %v for replay: %v", id, i, cmd, err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (transformer *Transformer) TransformAll(ctx context.Context) error {
```

Comment:
I've resolved two of my own comments inside this function. As we talked about, I think the contents of this function needs to be rearranged into the transform chain, but for now, this is fine to bootstrap the rest of the system, and we can revisit this when the true ControlFlowGenerator gets added.